File: CPlusPlus/OCP/Before.cpp
// NotificationService.cpp
#include <iostream>
#include <string>
#include <stdexcept>

class NotificationService {
public:
    void SendNotification(const std::string& message, const std::string& notificationType) {
        if (notificationType == "Email") {
            SendEmailNotification(message);
        }
        else if (notificationType == "SMS") {
            SendSmsNotification(message);
        }
        else if (notificationType == "PushNotification") {
            SendPushNotification(message);
        }
        else {
            throw std::invalid_argument("Notification type " + notificationType + " is not supported.");
        }
    }

private:
    void SendEmailNotification(const std::string& message) {
        std::cout << "Sending email notification: " << message << std::endl;
    }

    void SendSmsNotification(const std::string& message) {
        std::cout << "Sending SMS notification: " << message << std::endl;
    }

    void SendPushNotification(const std::string& message) {
        std::cout << "Sending push notification: " << message << std::endl;
    }
};

File: CSharp/OCP/Before.cs
public class NotificationService
{
    public void SendNotification(string message, string notificationType)
    {

        if (notificationType == "Email")
        {
            SendEmailNotification(message);
        }
        else if (notificationType == "SMS")
        {
            SendSmsNotification(message);
        }
        else if (notificationType == "PushNotification")
        {
            SendPushNotification(message);
        }
        else
        {
            throw new NotSupportedException($"Notification type {notificationType} is not supported.");
        }
    }

    private void SendEmailNotification(string message)
    {
        Console.WriteLine($"Sending email notification: {message}");
    }
    
    private void SendSmsNotification(string message)
    {
        Console.WriteLine($"Sending SMS notification: {message}");
    }

    private void SendPushNotification(string message)
    {
        Console.WriteLine($"Sending push notification: {message}");
    }
    
}